app/services/update_activity_as_fund.rb

```diff
@@ -22,7 +19,7 @@ def call
     activity.accountable_organisation_reference = "GB-GOV-13"
     activity.accountable_organisation_type = "10"
 
-    activity.save!
+    activity.save if activity.valid?(:parent_step)
```

Comment:
I think validation contexts would be nicer and a fraction more efficient here, it would look like this:

`activity.save(context: :parent_step)`

I am wondering about if this save fails for some unknown reason we return the activity regardless, are we okay with that or would a noisey `save!` give us more to go on to debug? What would the user see?

Reply:
I'll test how `activity.save(context: :parent_step)` fits in, I hadn't thought of that 👍 

The main concern I had here was that the error messages for validations are bubbled back up to the controller, so I removed the bang. I believe it catches unexpected problems and throws a 500 instead - something we shouldn't do by default as it is a jarring user experience.

On another look it seems that `ActivityFormController#update` will render the next step whatever happens:

```
render_wizard @activity, context: :"#{step}_step"
```

Given `update_form_state` guards against changing the form_state unless the activity is valid, perhaps we could guard against thie update taking us to the wrong step (without errors) by making the following change so we can lose the bangs:

```
render_wizard @activity, context: :"#{@activity.form_state}_step"
```

As mentioned the whole action is in dire need of some TLC. I'd be interested in finding a way to unblock this work and set ourselves up to refactor next.

Reply:
On further reading `render_wizard` calls `.save` on the object with the given context. Although this highlights that there are 3/4 saves of the same record which is not efficient, the form will not go to the next step unless it's valid. This means we are safe to remove the `!`, I think.

Tests are green when changing to `activity.save(context: :parent_step)` 👍